### PR TITLE
Implemented fix for #6006. Mime-type support

### PR DESCRIPTION
**Pull request by gupbeheer**
Technical committee: @TiFu @taxpon  @sebastianhaas  @kenisteward  @Vrolijkx 

### PR checklist

- [X] Read the [contribution guidelines](https://github.com/swagger-api/swagger-codegen/blob/master/CONTRIBUTING.md).
- [X ] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [X] Filed the PR against the correct branch: `3.0.0` branch for changes related to OpenAPI spec 3.0. Default: `master`.
- [X] Copied the [technical committee](https://github.com/swagger-api/swagger-codegen/#swagger-codegen-technical-committee) to review the pull request if your PR is targeting a particular programming langauge.

### Description of the PR
Fix for #6006 and #6454. Since the fix done in #6454 has been removed.
I tested using the Angular-v2 and v4 scripts.
This fix is needed to support rest services that use mime-types for versioning.
The configuration has been expanded with methods that determine the correct Accepts and Content-Type (for non-form posts) mime-type and can be overridden if needed when creating a configuration. (If multiple versions are supported by the rest service, the correct filtering can be done there)

There is one thing I am not completely satisfied with and that is that all generated operations have a 'consumes' block. The reason behind that is that there should be no duplicate code in the 'api.service.mustache'. But I could not find a Mustache section that is equal to '#bodyParam' and '#hasFormParams' together.

### CODE REVIEW

File: modules/swagger-codegen/src/main/resources/typescript-angular/configuration.mustache [hunk 1]
```diff
@@ -23,4 +23,53 @@ export class Configuration {
...
+     * If no content type is found return the first found type if the contentTypes is not empty
+     * @param {string[]} accepts - the array of content types that are available for selection.
+     * @returns {string} the selected content-type or <code>undefined</code> if no selection could be made.
+     */
+    public selectHeaderAccept(accepts: string[]): string  {
+        if (accepts.length == 0)
+            return undefined;
+
+        let type = accepts.find(x => this.isJsonMime(x));
+        if (type === undefined)
```
Comment (macjohnny) [inline modules/swagger-codegen/src/main/resources/typescript-angular/configuration.mustache]: nice, this should fix https://github.com/swagger-api/swagger-codegen/issues/6588

File: modules/swagger-codegen/src/main/resources/typescript-angular/configuration.mustache [hunk 2]
```diff
@@ -23,4 +23,53 @@ export class Configuration {
...
+     * If no content type is found return the first found type if the contentTypes is not empty
+     * @param {string[]} contentTypes - the array of content types that are available for selection
+     * @returns {string} the selected content-type or <code>undefined</code> if no selection could be made.
+     */
+    public selectHeaderContentType (contentTypes: string[]): string {
+        if (contentTypes.length == 0)
+            return undefined;
+
+        let type = contentTypes.find(x => this.isJsonMime(x));
+        if (type === undefined)
```
Comment (macjohnny) [inline modules/swagger-codegen/src/main/resources/typescript-angular/configuration.mustache]: I suggest using block braces `{` and `}` as usually linting rules discourage if statements without braces

File: modules/swagger-codegen/src/main/resources/typescript-angular/configuration.mustache [hunk 3]
```diff
@@ -23,4 +23,53 @@ export class Configuration {
...
+     * JSON MIME examples:
+     *   application/json
+     *   application/json; charset=UTF8
+     *   APPLICATION/JSON
+     *   application/vnd.company+json
+     * @param {string} mime - MIME (Multipurpose Internet Mail Extensions)
+     * @return {boolean} True if the given MIME is JSON, false otherwise.
+     */
+    public isJsonMime(mime: string): boolean {
+        let jsonMimeRegEx = new RegExp("^(application\/json|[^;/ \t]+\/[^;/ \t]+[+]json)[ \t]*(;.*)?$");
```
Comment (macjohnny) [inline modules/swagger-codegen/src/main/resources/typescript-angular/configuration.mustache]: how about also supporting json patch mime types?

the latest version of the `isJsonMime()` method before it was removed can be found here:
https://github.com/swagger-api/swagger-codegen/pull/6574/files#diff-3ab1141e07abf8762d6e33d15c9d1f0eL63

```
    public isJsonMime(mime: string): boolean {		
        const jsonMime: RegExp = new RegExp('^(application\/json|[^;/ \t]+\/[^;/ \t]+[+]json)[ \t]*(;.*)?$', 'i');		
        return mime != null && (jsonMime.test(mime) || mime.toLowerCase() === 'application/json-patch+json');		
    }
```

**Comment by wing328** [comment]
cc @TiFu @taxpon @sebastianhaas @kenisteward @Vrolijkx

**Comment by wing328** [comment]
I'll merge it this weekend if there's no further feedback/question.

**Comment by gupbeheer** [comment]
I thought I could fix the merge conflicts by a new build. But that does not seem to be the way.@wing328 if there is anything I can/should do to get this PR merged, please let me know.

**Comment by TiFu** [comment]
You can fix the merge requests by rebasing your branch on the current master.

See https://help.github.com/articles/syncing-a-fork/ for updating your repos master branch to this repos master branch.

Then you should be able to do a rebase (https://git-scm.com/docs/git-rebase) typescript_angular_mimetype_support onto master.

**Comment by wing328** [comment]
I just resolved the merge conflicts. Will merge later ...

**Comment by gamb79** [comment]
@wing328 Thank you! You're the best!

**Pull request merged**